### ISSUE TITLE
Intellisense not working properly with bound functions

### ISSUE THREAD

**Issue by aeschli** [Microsoft/TypeScript#6713]
_From @moci on January 28, 2016 10:43_

In general intellisense works just fine, see the following trivial example:
![good](https://cloud.githubusercontent.com/assets/464954/12641927/deaa630e-c5b2-11e5-9883-3cb843e72e4b.png)]
Intellisense suggests both the 2 members of "this", as expected.

However when using a bound function, the handler of the setTimeout method in this case, intellisense is unable to scope "this" to the class. It suggests nothing:
![bad](https://cloud.githubusercontent.com/assets/464954/12642068/caa24358-c5b3-11e5-9841-b9801fb8ef77.png)

We can give a hint to intellisense, afterwards it comes up with the correct suggestion:
![afterhelp](https://cloud.githubusercontent.com/assets/464954/12642086/e7d0af8c-c5b3-11e5-8fa3-f8738dd4d75e.png)

The real issue is when refactoring, find all references does not properly work:
![references](https://cloud.githubusercontent.com/assets/464954/12642137/2a2acbb0-c5b4-11e5-859f-1efce1dc0448.png)

_Copied from original issue: Microsoft/vscode#2485_

**Comment by techird** [comment]
`Function.prototy.bind` can not copy the type of the origin function. Why not just just arraw functions instead?

```ts
doDelay() {
    setTimeout(() => this.someProperty);
}
```

**Comment by jesseschalken** [comment]
`this` only has the right type in class methods and arrow functions within class methods. Everywhere else, `this` is `any`.

#3694 is the issue for typing `this` in more situations, although I'm not sure if it addresses your usage specifically.

**Comment by moci** [comment]
Thanks for the guidance! Sorry for creating an unnecessary issue ticket.

In the actual code I *was* using lambda functions but was not convinced about the correct context of `this`. So I began applying the `var self = this` pattern but was not sure that this was a good approach. Then I came upon `.bind(this)` and I noticed VSC not giving me the correct intellisense anymore. 

Thank you for pointing out that lambda's are the way to go!
[http://www.typescriptlang.org/Handbook#functions-lambdas-and-using-39this39](http://www.typescriptlang.org/Handbook#functions-lambdas-and-using-39this39)
